README.md
![Version 0.0.7](https://img.shields.io/badge/Version-0.0.7-green.svg)
![Perl](https://img.shields.io/badge/Perl-5.x-yellow.svg)
[![GPLv3 License](https://img.shields.io/badge/License-GPLv3-red.svg)](https://github.com/rezasp/joomscan/blob/master/LICENSE.md)
[![Twitter](https://img.shields.io/badge/Twitter-@OWASP_JoomScan-blue.svg)](http://twitter.com/OWASP_JoomScan)
[![Leader](https://img.shields.io/badge/Twitter-@rezesp-blue.svg)](http://www.twitter.com/rezesp)
[![Leader](https://img.shields.io/badge/Twitter-@Ali_Razmjo0-blue.svg)](http://www.twitter.com/Ali_Razmjo0)
<br>
[![Black Hat Arsenal USA](https://rawgit.com/toolswatch/badges/master/arsenal/usa/2018.svg)](http://www.toolswatch.org/2018/05/black-hat-arsenal-usa-2018-the-w0w-lineup/)
[![Black Hat Arsenal ASIA](https://rawgit.com/toolswatch/badges/master/arsenal/asia/2018.svg)](http://www.toolswatch.org/2018/01/black-hat-arsenal-asia-2018-great-lineup/)


<img src="https://raw.githubusercontent.com/rezasp/Trash/master/joomscan.png" width="200"><img src="https://raw.githubusercontent.com/rezasp/Trash/master/owasp.png" width="500">

======

OWASP JoomScan Project
======

OWASP Joomla! Vulnerability Scanner (JoomScan) is an open source project, developed with the aim of automating the task of vulnerability detection and reliability assurance in Joomla CMS deployments. Implemented in Perl, this tool enables seamless and effortless scanning of Joomla installations, while leaving a minimal footprint with its lightweight and modular architecture. It not only detects known offensive vulnerabilities, but also is able to detect many misconfigurations and admin-level shortcomings that can be exploited by adversaries to compromise the system. Furthermore, OWASP JoomScan provides a user-friendly interface and compiles the final reports in both text and HTML formats for ease of use and minimization of reporting overheads.
<br>
OWASP JoomScan is included in Kali Linux distributions.

### WHY OWASP JOOMSCAN  ?
Automated ...<br>
  *Version enumerator<br>
  *Vulnerability enumerator (based on version)<br>
  *Components enumerator (1209 most popular by default)<br>
  *Components vulnerability enumerator (based on version)(+1030 exploit)<br>
  *Firewall detector<br>
  *Reporting to Text & HTML output<br>
  *Finding common log files<br>
  *Finding common backup files<br>


# INSTALL

    git clone https://github.com/rezasp/joomscan.git
    cd joomscan
    perl joomscan.pl


# JOOMSCAN ARGUMENTS

    Usage:	joomscan.pl [options]

    --url | -u <URL>                |   The Joomla URL/domain to scan.
    --enumerate-components | -ec    |   Try to enumerate components.

    --cookie <String>               |   Set cookie.
    --user-agent | -a <user-agent>  |   Use the specified User-Agent.
    --random-agent | -r             |   Use a random User-Agent.
    --timeout <time-out>            |   set timeout.
    --about                         |   About Author
    --update                        |   Update to the latest version.
    --help | -h                     |   This help screen.
    --version                       |   Output the current version and exit.


# OWASP JOOMSCAN USAGE EXAMPLES

Do default checks...<br>
```perl joomscan.pl --url www.example.com```<br>
or<br>
```perl joomscan.pl -u www.example.com```
<br>
<br>
Enumerate installed components...<br>
```perl joomscan.pl --url www.example.com --enumerate-components```<br>
or<br>
```perl joomscan.pl -u www.example.com --ec```<br>
<br>

Set cookie<br>
```perl joomscan.pl --url www.example.com --cookie "test=demo;"```
<br><br>

Set user-agent<br>
```perl joomscan.pl --url www.example.com --user-agent "Googlebot/2.1 (+http://www.googlebot.com/bot.html)"```<br>
or<br>
```perl joomscan.pl -u www.example.com -a "Googlebot/2.1 (+http://www.googlebot.com/bot.html)"```<br>
<br><br>

Set random user-agent<br>
```perl joomscan.pl -u www.example.com --random-agent```<br>
or<br>
```perl joomscan.pl --url www.example.com -r```<br>
<br>

Set proxy<br>
```perl joomscan.pl --url www.example.com --proxy http://127.0.0.1:8080```<br>
or<br>
```perl joomscan.pl -u www.example.com --proxy https://127.0.0.1:443```<br>
<br><br>

Update Joomscan...<br>
```perl joomscan.pl --update```<br><br>


# OWASP PAGE

[https://www.owasp.org/index.php/Category:OWASP_Joomla_Vulnerability_Scanner_Project](https://www.owasp.org/index.php/Category:OWASP_Joomla_Vulnerability_Scanner_Project)

# GIT REPOSITORY

[https://github.com/rezasp/joomscan](https://github.com/rezasp/joomscan)

# ISSUES

[https://github.com/rezasp/joomscan/issues](https://github.com/rezasp/joomscan/issues)

# PROJECT LEADERS

*  Mohammad Reza Espargham           [ reza[dot]espargham[at]owasp[dot]org ]
*  Ali Razmjoo                                    [ ali[dot]razmjoo[at]owasp[dot]org ]


<br><br>
OWASP JoomScan introduction (Youtube)

[![OWASP JoomScan introduction](https://img.youtube.com/vi/Ik2CJ9LkuoI/0.jpg)](https://www.youtube.com/watch?v=Ik2CJ9LkuoI)

CHANGELOG.md
OWASP JoomScan 0.0.7 [Self Challenge]
============
* com_joomanager exploiter removed
* Added new module: Local File Disclosure vulnerability detector (Supports detection of [com_joomanager,s5_media_player,com_hdflvplayer,com_macgallery,com_cckjseblod,fsave,com_portfolio,com_picsell,captcha,com_rsfiles,com_addproperty,com_aceftp,com_jtagmembersdirectory,com_facegallery,com_docman,mod_dvfoldercontent,com_contushdvideoshare,com_jetext,com_product_modul,wddownload,com_community,com_download-monitor])
* Updated  module: Firewall Detector (supports detection of [CloudFlare, Incapsula, Shieldfy, Mod_Security and 28 other modules ])
* Added exploit for jckeditor
* Updated list of components
* A few enhancements

OWASP JoomScan 0.0.6 [#BHUSA]
============
* Updated vulnerability databases
* Added new module: Firewall Detector (supports detection of [CloudFlare, Incapsula, Shieldfy, Mod_Security])
* Added exploit for com_joomanager
* Updated list of common log paths
* A few enhancements

OWASP JoomScan 0.0.5 [KLOT]
============
* Update components database
* Bug fixed (updating module)
* Allow start from any path
* Update backup finder database
* Update report module
* Update validate target method 
* HTTPS improvements
* Fix issue #11 - Incorrect URL output for HTTPS site
* Fix issue #12 - Components scan output issues
* Fix issue #13 - Check a server is live or not!
* Fix issue #9 - Disable redirectable requests for components finder module
* A few enhancements

OWASP JoomScan 0.0.1 [Reborn]
============
* Initial release


core/header.pl
#!/usr/bin/perl

use Term::ANSIColor;
 
print color("YELLOW");
print q{
    ____  _____  _____  __  __  ___   ___    __    _  _ 
   (_  _)(  _  )(  _  )(  \/  )/ __) / __)  /__\  ( \( )
  .-_)(   )(_)(  )(_)(  )    ( \__ \( (__  /(__)\  )  ( 
  \____) (_____)(_____)(_/\/\_)(___/ \___)(__)(__)(_)\_)
};
   	print color("red") . "\t\t\t(1337.today)" . color("reset");
   	print "
   
    --=[". color("BLUE") . "OWASP JoomScan". color("reset") ."
    +---++---==[Version : "
   	. color("red"). "$version\n". color("reset") . "    +---++---==[Update Date : [". color("red") . "$update". color("reset") . "]
    +---++---==[Authors : ". color("red") . "$author". color("reset")."
    --=[Code name : ". color("red") . "$codename". color("reset")."\n    \@OWASP_JoomScan , \@rezesp , \@Ali_Razmjo0 , \@OWASP\n\n";
    	
if(!defined $ARGV[0]){
 
    print color("cyan");
    printf "\n   Usage: 
    	joomscan.pl <target>
   	joomscan.pl -u http://target.com/joomla
   
   
      Options: 
   	joomscan.pl --help\n\n";
    print color("reset");
    exit(1);
}
$cookie=1;
$proxy=1;
#Start help Function
sub help
{
	print color("cyan");
	printf "   

Help :

Usage:	$0 [options]

--url | -u <URL>                |   The Joomla URL/domain to scan.
--enumerate-components | -ec    |   Try to enumerate components.

--cookie <String>               |   Set cookie.
--user-agent | -a <User-Agent>  |   Use the specified User-Agent.
--random-agent | -r             |   Use a random User-Agent.
--timeout <Time-Out>            |   Set timeout.
--proxy=PROXY                   |   Use a proxy to connect to the target URL
           Proxy example: --proxy http://127.0.0.1:8080
                                  https://127.0.0.1:443
                                  socks://127.0.0.1:414
                                  
--about                         |   About Author
--update                        |   Update to the latest version.
--help | -h                     |   This help screen.
--version                       |   Output the current version and exit.

";
	print color("reset");
	exit(1);
}
sub about
{
	print color("cyan");
	printf "
   Author         :   $author
   Twitter        :   \@rezesp , \@Ali_Razmjo0
   Git repository :   https://github.com/rezasp/joomscan/
   Issues         :   https://github.com/rezasp/joomscan/issues
    \n\n";
	print color("reset");
	exit(1);
}
sub update
{
    do "$mepath/core/update.pl";
	print color("reset");
	exit(1);
}


GetOptions(
  'help|h' => sub { help(0) },
  'update' => sub { update(0) },
  'about' => sub { about(0) },
  'enumerate-components|ec'   => sub { $components = 1 },
  'random-agent|r'   => sub { $randomagent = 1 },
  'user-agent|a=s' => \$agent,
  'timeout=s' => \$timeout,
  'proxy=s' => \$proxy,
  'cookie=s' => \$cookie,
  'u|url=s' => \$target,
  'version' => sub { print "\n\nVersion : $version\n\n";exit; },

);
if($target !~ /\S/){
  print color("red");
  print "[+] No target specified!\n\n";
  print color("reset");
  exit (1);
}
if($target !~ /^https?:\/\//) { $target = "http://$target"; };

#End help Function
